awx/main/analytics/collectors.py

```diff
@@ -126,47 +124,24 @@ def projects_by_scm_type(since):
-    counts['error_jobs'] = models.UnifiedJob.objects.filter(status='error').count()   # Do we also want to include error, new, pending, waiting, running?
-    counts['new_jobs'] = models.UnifiedJob.objects.filter(status='new').count()       # Also, how much of this do we want `per instance`
-    counts['pending_jobs'] = models.UnifiedJob.objects.filter(status='pending').count()
-    counts['waiting_jobs'] = models.UnifiedJob.objects.filter(status='waiting').count()
-    counts['running_jobs'] = models.UnifiedJob.objects.filter(status='running').count()
-                                
-    # These will later be used to optimize the jobs_running and jobs_total python properties ^^
-    # jobs_running = models.UnifiedJob.objects.filter(execution_node=instance, status__in=('running', 'waiting',)).count()
-    # jobs_total = models.UnifiedJob.objects.filter(execution_node=instance).count()
-        
-    return counts
-    
-    
-@register('job_counts_instance')
-def job_counts_instance(since):
-    counts = {}
+    counts['status'] = dict(models.UnifiedJob.objects.values_list('status').annotate(Count('status')))
     for instance in models.Instance.objects.all():
         counts[instance.id] = {'uuid': instance.uuid,
-                               'jobs_total': instance.jobs_total,       # this is _all_ jobs run by that node
-                               'jobs_running': instance.jobs_running,   # this is jobs in running & waiting state
-                               'launch_type': {'manual': models.UnifiedJob.objects.filter(launch_type='manual').count(), # I can definitely condense this
-                                               'relaunch': models.UnifiedJob.objects.filter(launch_type='relaunch').count(),
-                                               'scheduled': models.UnifiedJob.objects.filter(launch_type='scheduled').count(),
-                                               'callback': models.UnifiedJob.objects.filter(launch_type='callback').count(),
-                                               'dependency': models.UnifiedJob.objects.filter(launch_type='dependency').count(),
-                                               'sync': models.UnifiedJob.objects.filter(launch_type='workflow').count(),
-                                               'scm': models.UnifiedJob.objects.filter(launch_type='scm').count()
-                                               }
-                               }
+                                'jobs_total': models.UnifiedJob.objects.filter(execution_node=instance.hostname, status__in=('running', 'waiting',)).count(),
+                                'jobs_running': models.UnifiedJob.objects.filter(execution_node=instance.hostname).count(),   # jobs in running & waiting state
+                                'launch_type': dict(models.UnifiedJob.objects.filter(execution_node=instance.hostname).values_list('launch_type').annotate(Count('launch_type')))
```

Comment:
👍

Reply:
If you've got e.g., 25 instances, this is 25 aggregate queries.  Is there a way to do this all as *one* query grouped by `(main_instance.hostname, main_unifiedjob.launch_type)`?